dot/parachain/prospective-parachains/active_leaves_update.go

```diff
@@ -0,0 +1,388 @@
+
+			scope, err := newScopeWithAncestors(
+				relayChainBlockInfo{
+					Hash:        blockInfo.Hash,
+					Number:      blockInfo.Number,
+					StorageRoot: blockInfo.StorageRoot,
+				},
+				constraints,
+				compactPending,
+				uint(maxBackableChainLen),
+				ancestryBlockInfo,
+			)
+			if err != nil {
+				logger.Warnf(
+					"relay chain ancestors have wrong order, "+
+						"para id=%d, max backable=%d, ancestry=%s, leaf=%s, err=%s",
+					paraID,
+					maxBackableChainLen,
+					strings.Join(ancestorsHashes, ","),
+					hash,
+					err.Error(),
+				)
+				continue
+			}
+
+			logger.Tracef(
+				"creating fragment chain, "+
+					"relay parent=%s, min relay parent=%s, max backable=%d, para id=%d, ancestors=%s",
+				hash,
+				scope.earliestRelayParent().Number,
+				maxBackableChainLen,
+				paraID,
+				strings.Join(ancestorsHashes, ","),
+			)
+
+			numOfPendingCandidates := pendingAvailabilityStorage.len()
+			// Init the fragment chain with the pending availability candidates.
+			chain := newFragmentChain(scope, pendingAvailabilityStorage)
+			if chain.bestChainLen() < numOfPendingCandidates {
+				logger.Warnf(
+					"not all pending availability candidates could be introduced, "+
+						"expected=%d, actual=%d, para id=%d, relay parent=%d",
+					chain.bestChainLen(),
+					numOfPendingCandidates,
+					paraID,
+					hash,
+				)
+			}
+
+			// If we know the previous fragment chain, use that for further populating the fragment
+			// chain.
+
+			if prevFragmentChains != nil {
+				if prev, ok := prevFragmentChains[paraID]; ok {
+					chain.populateFromPrevious(prev)
+				}
+			}
+
+			logger.Tracef("populated fragment chain with %d candidates: %v, relay parent=%s, para id=%d",
+				chain.bestChainLen(), chain.bestChainVec(), hash, paraID)
+
+			unconnectedHashes := make([]string, 0)
+			for hash := range chain.unconnected.byCandidateHash {
+				unconnectedHashes = append(unconnectedHashes, hash.String())
+			}
+
+			logger.Tracef("potential candidate storage for para: %v, relay parent=%s, para id=%d",
+				strings.Join(unconnectedHashes, ","), hash, paraID)
+
+			fragmentChains[paraID] = chain
+		}
+
+		pp.view.perRelayParent[hash] = &relayParentData{
+			fragmentChains: fragmentChains,
+		}
+
+		pp.view.activeLeaves[hash] = true
+		pp.view.implicitView.ActivateLeafFromProspectiveParachains(blockInfo, ancestry)
+	}
+
+	for _, deactivated := range msg.Deactivated {
+		delete(pp.view.activeLeaves, deactivated)
+		pp.view.implicitView.DeactivateLeaf(deactivated)
+	}
+
+	// keep in our state only the relay parents that are still active
+	// under implicitly view
+	allowedRelayParents := pp.view.implicitView.AllAllowedRelayParents()
+
+	rpToDelete := make([]common.Hash, 0)
+	for rp := range pp.view.perRelayParent {
+		if !slices.Contains(allowedRelayParents, rp) {
+			rpToDelete = append(rpToDelete, rp)
+		}
+	}
+
+	for _, rp := range rpToDelete {
+		delete(pp.view.perRelayParent, rp)
+	}
```

Comment:
It's safe to delete from maps during iteration.
```suggestion
	for rp := range pp.view.perRelayParent {
		if !slices.Contains(allowedRelayParents, rp) {
		        delete(pp.view.perRelayParent, rp)
		}
	}
```